### ISSUE TITLE
Lazy values

### ISSUE THREAD

**Issue by kayceesrk** [ocamllabs/ocaml-multicore#188]
Lazy values are deferred computations (thunks) that can be forced to compute the result. The implementation of lazy values for multicore need to take into account the presence of concurrently running mutators and the concurrent GC. 

In trunk, OCaml says forcing thunks is not a thread-safe operation. We make the semantics more precise and say that concurrently forcing a thunk would raise `Undefined` exception. The interaction with the concurrent GC thread is more interesting. `CamlInternalLazy` uses `Obj.set_tag` to update the header from `Lazy_tag` to `Forward_tag` when it forces thunks. However, `set_tag` is unsafe under multicore and there is a [proposal](https://github.com/ocaml/ocaml/pull/1725) to deprecate it. 

We enrich the layout of lazy values with an extra word to make it 3 words in total including the header. It has the following layout:

```ocaml
| Hd(Lazy_tag) | status | content |
```

The status field is a value of type `status`:

```ocaml
type status = Forced | Unforced | Forcing
```

If the status field is `Unforced` then the content is a closure and if the status is `Forced` then the content is the result of forcing the thunk. The status `Forcing` is used when the lazy is value is currently being forced by some domain. Properly synchronised non-recursive lazy forcing will never observe `Forcing`. 

The implementation of `force` is:

```ocaml
exception Undefined

let rec force (lv : 'a lazy_t) = 
  let s = get_status lv in
  let x = Obj.repr lv in 
  match s with
  | Forced -> (Obj.obj (Obj.field x 1) : 'a) (* value *)
  | Forcing -> raise Undefined
  | Unforced -> 
    if Obj.compare_and_swap_field x 0 (Obj.repr Unforced) (Obj.repr Forcing) then begin
      (* successfully updated status to forcing *)
      let thunk = (Obj.obj (Obj.field x 1) : unit -> 'a) in
      let result = thunk () in
      Obj.set_field x 1 (Obj.repr result);
      Obj.set_field x 0 (Obj.repr Forced)
    end else (* someone else beat us to the CAS or recursive forcing. *)
      force lv
```

We no longer modify the header field as part of the forcing and hence forcing a lazy is safe wrt to the GC.

**Comment by kayceesrk** [comment]
We can make lazy thread-safe by using an [`IVar`](http://cml.cs.uchicago.edu/pages/sync-var.html#SIG:SYNC_VAR.ivar:TY) that records the other domains that are concurrently forcing. This extensions and others should be implemented after the initial multicore release.

**Comment by Gbury** [comment]
Thanks. However, maybe it's me being slow, but I don't see where the case of a short circuited thunk is handled in the code of `force` you showed, or is it handled somewhere else ?

**Comment by kayceesrk** [comment]
Short-circuiting happens in the GC which is not showed here. The GC code for short-circuiting remains the same except for how the GC determines whether a lazy is forwarded.

**Comment by Gbury** [comment]
Yeah, that part I get, but when you force a thunk that has been short circuited by the GC, its tag is not a lazy tag, so what does `get_status` at the beginning of `force` do, since the layout of the block can be anything (the forced value, if I'm not mistaken) ? Or is there a wrapper around the code you showed that handles this case ?
For reference in the current (though not technically used according to the comments), one can see this case handled at https://github.com/ocaml/ocaml/blob/be3730d8a32fe4675c255351bea3563d8a2d4907/stdlib/camlinternalLazy.ml#L56 .

**Comment by kayceesrk** [comment]
Thanks @Gbury. I see your point now. I've been overlooking some of the subtleties. In order to handle short-circuited block, we need to add an extra check:

```ocaml
exception Undefined

let rec force (lv : 'a lazy_t) = 
  let x = Obj.repr lv in
  if Obj.tag x <> Obj.lazy_tag then (Obj.obj x : 'arg)
  else begin 
    match (Obj.obj (Obj.field x 0): status) with
    | Forced -> (Obj.obj (Obj.field x 1) : 'a) (* value *)
    | Forcing -> raise Undefined
    | Unforced -> 
      if Obj.compare_and_swap_field x 0 (Obj.repr Unforced) (Obj.repr Forcing) then begin
        (* successfully updated status to forcing *)
        let thunk = (Obj.obj (Obj.field x 1) : unit -> 'a) in
        let result = thunk () in
        Obj.set_field x 1 (Obj.repr result);
        Obj.set_field x 0 (Obj.repr Forced)
      end else (* someone else beat us to the CAS or recursive forcing. *)
        force lv
  end
```

I think the correctness of this code is quite subtle since a GC shouldn't run between the check for whether the tag of x is not `Obj.lazy_tag` and the compare and swap. Otherwise, there is a risk that we end up successfully compare and swapping on the short-circuited value.

**Comment by lpw25** [comment]
I can't remember whether this is currently the case, but I'm not sure that GC roots on the stack get short-circuited. This would make the code more obviously safe since `x` can't just change what it points to.

**Comment by kayceesrk** [comment]
Stack values do get short-circuited:

```bash
$ cat test.ml
let string_of_opt_int = function
  | None -> "None"
  | Some v -> "Some " ^ (string_of_int v)

let _ =
  let l = lazy (print_string "Forced\n"; Some 10) in
  Printf.printf "Tag is %d\n" (Obj.tag (Obj.repr l));
  Printf.printf "Value is %s\n" @@ string_of_opt_int @@ Lazy.force l;
  Printf.printf "Tag is %d\n" (Obj.tag (Obj.repr l));
  Gc.minor ();
  Printf.printf "After GC\n";
  Printf.printf "Value is %s\n" @@ string_of_opt_int @@ Lazy.force l;
  Printf.printf "Tag is %d\n" (Obj.tag (Obj.repr l));
$ ocaml test.ml
Tag is 246
Forced
Value is Some 10
Tag is 250
After GC
Value is Some 10
Tag is 0
```

246 is `Obj.lazy_tag` & 250 is `Obj.forward_tag`.

**Comment by lpw25** [comment]
I which case there should definitely be a compiler primitive to do this rather than OCaml code. Then we can ensure that we don't have any issues without relying on the rules for GC safe-points. For example,
we could mark `x` as `Addr` to ensure that a GC safe-point cannot be inserted in the wrong place.

**Comment by kayceesrk** [comment]
An alternative design is to turn `Obj.set_tag` to a compare-and-swap and marking Lazy tag objects to use compare-and-swap to update the header. This design would avoid the extra word in the lazy object at the cost of slowing down marking and short-circuiting lazy. We can benchmark the alternatives to quantify the costs. Thoughts? @stedolan.

**Comment by stedolan** [comment]
That's a good idea! We already branch on the tag during marking, so I don't imagine an extra branch on Lazy_tag will cost much. Should be easy to benchmark too.

I still think we should remove Obj.set_tag, though, as it's still unsafe in general. We'll need a new primitive for setting the tags of Lazy values, that doesn't work on any other tag.

**Comment by kayceesrk** [comment]
Here is a different solution to the lazy problem that doesn't use an extra word per object: https://github.com/ocamllabs/ocaml-multicore/pull/226.

**Comment by kayceesrk** [comment]
Fixed by #226.
